If there is a missing request in the XHR mock, print a message to the error console.

shane-tomlinson opened the pull request:
@ozten - you wanna review this?

fixes #2759

Commit:
If there is a missing request in the XHR mock, print a message to the error console.

ozten commented:
Will do, but I won't get to it anytime soon. Thanks for busting this out so quickly.

ozten commented on resources/static/test/mocks/xhr.js:
```diff
@@ -221,10 +223,14 @@ BrowserID.Mocks.xhr = (function() {
       this.requests[request.url] = request;
 
+      if (!(responseKey in xhr.responses)) {
+        console.error("request for '" + request.url + "' is not mocked in resources/static/test/cases/xhr.js");
```
Very cool! This will be helpful for spotting updates that break xhr.

ozten commented on resources/static/test/mocks/xhr.js:
```diff
@@ -221,10 +223,14 @@ BrowserID.Mocks.xhr = (function() {
         response(request.success);
       }
-      else if (!(typeofResponse == "number" || typeofResponse == "undefined")) {
+      else if (!(typeofResponse === "number" || typeofResponse === "undefined")) {
```
Nice cleanup!

ozten commented:
Running `npm test` locally, I get lots of output... is that expected? If so, maybe have a way to hint to the tests to ignore an expected xhr failure?

```
request for '/wsapi/email_addition_status?email=testuser%40testuser.com' is not mocked in resources/static/test/cases/xhr.js
request for '/wsapi/add_email_with_assertion' is not mocked in resources/static/test/cases/xhr.js
request for '/wsapi/email_for_token?token=token' is not mocked in resources/static/test/cases/xhr.js
request for '/wsapi/update_password' is not mocked in resources/static/test/cases/xhr.js
request for '/wsapi/user_creation_status?email=registered%40testuser.com' is not mocked in resources/static/test/cases/xhr.js
request for '/wsapi/email_for_token?token=token' is not mocked in resources/static/test/cases/xhr.js
request for '/wsapi/email_for_token?token=token' is not mocked in resources/static/test/cases/xhr.js
request for '/wsapi/address_info?email=registered%40testuser.com' is not mocked in resources/static/test/cases/xhr.js
request for '/wsapi/email_for_token?token=token' is not mocked in resources/static/test/cases/xhr.js
request for '/wsapi/have_email?email=registered' is not mocked in resources/static/test/cases/xhr.js
request for '/wsapi/email_for_token?token=token' is not mocked in resources/static/test/cases/xhr.js
request for '/wsapi/email_for_token?token=token' is not mocked in resources/static/test/cases/xhr.js
request for '/wsapi/password_reset_status?email=registered%40testuser.com' is not mocked in resources/static/test/cases/xhr.js
request for '/wsapi/password_reset_status?email=registered%40testuser.com' is not mocked in resources/static/test/cases/xhr.js
request for '/wsapi/address_info?email=unregistered%40testuser.com' is not mocked in resources/static/test/cases/xhr.js
request for '/wsapi/user_creation_status?email=registered%40testuser.com' is not mocked in resources/static/test/cases/xhr.js
request for '/wsapi/user_creation_status?email=%20%20registered%40testuser.com%20%20' is not mocked in resources/static/test/cases/xhr.js
request for '/wsapi/list_emails' is not mocked in resources/static/test/cases/xhr.js
request for '/wsapi/list_emails' is not mocked in resources/static/test/cases/xhr.js
request for '/wsapi/list_emails' is not mocked in resources/static/test/cases/xhr.js
request for '/wsapi/address_info?email=unregistered%40testuser.com' is not mocked in resources/static/test/cases/xhr.js
request for '/wsapi/address_info?email=registered%40testuser.com' is not mocked in resources/static/test/cases/xhr.js
currently unable to focus elements, focus check skipped - try focusing the unit test page
currently unable to focus elements, focus check skipped - try focusing the unit test page
currently unable to focus elements, focus check skipped - try focusing the unit test page
request for '/wsapi/user_creation_status?email=registered%40testuser.com' is not mocked in resources/static/test/cases/xhr.js
request for '/wsapi/user_creation_status?email=registered%40testuser.com' is not mocked in resources/static/test/cases/xhr.js
currently unable to focus elements, focus check skipped - try focusing the unit test page
request for '/wsapi/address_info?email=registered%40testuser.com' is not mocked in resources/static/test/cases/xhr.js
request for '/wsapi/add_email_with_assertion' is not mocked in resources/static/test/cases/xhr.js
```

shane-tomlinson commented:
Thanks @ozten, I'll update. I know there were some that I intentionally skipped adding because I knew the behavior was to call the onFailure callback, which is what I wanted in those circumstances.

I will fix this.

Commit:
Ensure that all expected XHR requests exist in the XHR mock.

ajaxError requests are the only type allowed to be undefined.

This pointed out a small problem in pages/js/signin.js where a new email address could be submitted without first being trimmed.

* Make sure all expected XHR requests exist.
* Generalize names (xhrError=>ajaxError, registered=>registered@testuser.com).
* Make sure "done" callbacks are called when everything is complete.
* Make sure email addresses are trimmed in pages/js/signin.js

shane-tomlinson commented on resources/static/pages/js/signin.js:
```diff
@@ -124,7 +124,7 @@ BrowserID.signIn = (function() {
   function signUpSubmit(oncomplete) {
     /*jshint validthis: true*/
-    var email = dom.getInner("#email"),
+    var email = helpers.getAndValidateEmail("#email"),
```
This line is from PR #2782.

shane-tomlinson commented on resources/static/pages/js/page_helpers.js:
```diff
@@ -95,8 +95,8 @@ BrowserID.PageHelpers = (function() {
       userValidationComplete(status);
+      onComplete && onComplete();
     });
-    onComplete && onComplete();
```
onComplete should have been called after the poll function completed.

shane-tomlinson commented on resources/static/test/cases/pages/js/manage_account.js:
```diff
@@ -48,22 +48,24 @@
       $("#old_password").val(oldPass);
       $("#new_password").val(newPass);
 
+      if (xhrResult) xhr.useResult(xhrResult);
```
This change is so we can set the XHR result of the changePassword request without affecting the requests that occur during createController.

shane-tomlinson commented on resources/static/test/cases/pages/js/signin.js:
```diff
@@ -139,7 +139,7 @@
 
   asyncTest("address with XHR error - show error screen", function() {
-    xhr.useResult("xhrError");
+    xhr.useResult("ajaxError");
```
Use the standard ajaxError instead of xhrError.

ozten commented:
Awesome. Now seeing only

```
--
```

[12:24:05.503] request for 'get /wsapi/address_info?email=unregistered3%40testuser.com unknown_secondary' is not mocked in resources/static/test/cases/xhr.js

as expected. Great work!
